Panic when moving to neoq_dead_jobs

Thank you for the patch for https://github.com/acaloiaro/neoq/issues/85, however this gets me to the next problems. 

First, when a failed job exceeds the max retries, it is to be moved to the `neoq_dead_jobs` table, however a panic is being generated there:

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x18 pc=0x1a53181]

goroutine 118 [running]:
github.com/acaloiaro/neoq/backends/postgres.(*PgBackend).moveToDeadQueue(0xc000774150?, {0x20bd150, 0xc000774150}, {0x20c4300, 0xc0000126a8}, 0xc0001f5680, {0x0, 0x0})
	/Users/billmatlock/go/pkg/mod/github.com/acaloiaro/neoq@v0.36.0/backends/postgres/postgres_backend.go:442 +0xe1
github.com/acaloiaro/neoq/backends/postgres.(*PgBackend).updateJob(0xc0003f4000, {0x20bd150, 0xc000774150}, {0x0, 0x0})
	/Users/billmatlock/go/pkg/mod/github.com/acaloiaro/neoq@v0.36.0/backends/postgres/postgres_backend.go:484 +0x2c5
github.com/acaloiaro/neoq/backends/postgres.(*PgBackend).handleJob(0xc0003f4000, {0x20bd188, 0xc00039c370}, {0x28a0a38, 0x1}, {0xc000412150, 0x8, 0x6fc23ac00, 0x0, {0x1dc5b27, ...}})
	/Users/billmatlock/go/pkg/mod/github.com/acaloiaro/neoq@v0.36.0/backends/postgres/postgres_backend.go:723 +0x4ce
github.com/acaloiaro/neoq/backends/postgres.(*PgBackend).start.func2()
	/Users/billmatlock/go/pkg/mod/github.com/acaloiaro/neoq@v0.36.0/backends/postgres/postgres_backend.go:539 +0x1db
created by github.com/acaloiaro/neoq/backends/postgres.(*PgBackend).start in goroutine 5
	/Users/billmatlock/go/pkg/mod/github.com/acaloiaro/neoq@v0.36.0/backends/postgres/postgres_backend.go:532 +0x353
```

I believe this is due to the fact that `jobErr` is not being populated, causing the `nil pointer dereference`. However the error message is also in the job struct so maybe the fix would be to drop the `jobErr` param and set the `error` field to the value of `j.Error` ? 

In `/backends/postgres/postgres_backend.go moveToDeadQueue()`
Go from: 
```golang
_, err = tx.Exec(ctx, `INSERT INTO neoq_dead_jobs(id, queue, fingerprint, payload, retries, max_retries, error, deadline)
	VALUES ($1, $2, $3, $4, $5, $6, $7, $8)`,
	j.ID, j.Queue, j.Fingerprint, j.Payload, j.Retries, j.MaxRetries, jobErr.Error(), j.Deadline)
```

To:
```golang
_, err = tx.Exec(ctx, `INSERT INTO neoq_dead_jobs(id, queue, fingerprint, payload, retries, max_retries, error, deadline)
	VALUES ($1, $2, $3, $4, $5, $6, $7, $8)`,
	j.ID, j.Queue, j.Fingerprint, j.Payload, j.Retries, j.MaxRetries, j.Error, j.Deadline)
```

Secondly, if I patch the above error as described to get through the panic, the next time the failed job is ran, it is being ran on a different queue (handler) than what is assigned to it. Here is a snippet of my testing log showing that the first time it is getting handled by the correct handler (`nqTestHandlerWithErr()`) but the second time it is not (`nqTestHandler()`), but I do not know if my changes above are causing this or not, so I did not open a separate issue:

```
=== RUN   TestNeoq/TestAddJobs
2023/10/13 11:10:22 nqTestHandler()
2023/10/13 11:10:22 got job id: 1 queue: test_now messsage: [test_now] this is an instant job
2023/10/13 11:10:22 waiting 10 seconds for delayed job to run...
2023/10/13 11:10:27 nqTestHandler()
2023/10/13 11:10:27 got job id: 2 queue: test_delayed messsage: [test_delayed] this is a delayed job
2023/10/13 11:10:32 waiting 3 seconds for failed job to run...
2023/10/13 11:10:32 nqTestHandlerWithErr()
2023/10/13 11:10:32 got job id: 3 queue: test_error messsage: [test_error] this is a job with errors
time=2023-10-13T11:10:32.495-04:00 level=ERROR msg="job failed" job_error="job failed to process: this is a test error"
2023/10/13 11:10:35 waiting 30 seconds for failed job to get retried...
2023/10/13 11:10:48 nqTestHandler() <<<<< should be `nqTestHandlerWithErr()`
2023/10/13 11:10:48 got job id: 3 queue: test_error messsage: [test_error] this is a job with errors
```

And here are the test handlers:
```golang
var (
	ErrTest = errors.New("this is a test error")
)

func nqTestHandler(db *database.DB) handler.Func {
	return func(ctx context.Context) (err error) {
		var j *jobs.Job
		j, err = jobs.FromContext(ctx)
		if err != nil {
			return
		}
		log.Println("nqTestHandler()")
		log.Println("got job id:", j.ID, "queue:", j.Queue, "messsage:", j.Payload["message"])
		return
	}
}

func nqTestHandlerWithErr() handler.Func {
	return func(ctx context.Context) (err error) {
		j, err := jobs.FromContext(ctx)
		if err != nil {
			log.Printf("ERROR in nqTestHandlerWithErr: %v", err)
		}
		log.Println("nqTestHandlerWithErr()")
		log.Println("got job id:", j.ID, "queue:", j.Queue, "messsage:", j.Payload["message"])
		return ErrTest
	}
}
```

Thank you!

--- a/backends/postgres/postgres_backend.go
+++ b/backends/postgres/postgres_backend.go
@@ -31,6 +31,12 @@ import (
 var migrationsFS embed.FS
 
 const (
+	JobQuery = `SELECT id,fingerprint,queue,status,deadline,payload,retries,max_retries,run_after,ran_at,created_at,error
+					FROM neoq_jobs
+					WHERE id = $1
+					AND status NOT IN ('processed')
+					FOR UPDATE SKIP LOCKED
+					LIMIT 1`
 	PendingJobIDQuery = `SELECT id
 					FROM neoq_jobs
 					WHERE queue = $1
@@ -38,14 +44,7 @@ const (
 					AND run_after <= NOW()
 					FOR UPDATE SKIP LOCKED
 					LIMIT 1`
-	PendingJobQuery = `SELECT id,fingerprint,queue,status,deadline,payload,retries,max_retries,run_after,ran_at,created_at,error
-					FROM neoq_jobs
-					WHERE id = $1
-					AND status NOT IN ('processed')
-					AND run_after <= NOW()
-					FOR UPDATE SKIP LOCKED
-					LIMIT 1`
-	FutureJobQuery = `SELECT id,run_after
+	FutureJobQuery = `SELECT id,fingerprint,queue,status,deadline,payload,retries,max_retries,run_after,ran_at,created_at,error
 					FROM neoq_jobs
 					WHERE queue = $1
 					AND status NOT IN ('processed')
@@ -75,7 +74,7 @@ type PgBackend struct {
 	cron        *cron.Cron
 	mu          *sync.RWMutex // mutex to protect mutating state on a pgWorker
 	pool        *pgxpool.Pool
-	futureJobs  map[string]time.Time       // map of future job IDs to their due time
+	futureJobs  map[string]*jobs.Job       // map of future job IDs to the corresponding job record
 	handlers    map[string]handler.Handler // a map of queue names to queue handlers
 	cancelFuncs []context.CancelFunc       // A collection of cancel functions to be called upon Shutdown()
 }
@@ -112,7 +111,7 @@ func Backend(ctx context.Context, opts ...neoq.ConfigOption) (pb neoq.Neoq, err 
 		mu:          &sync.RWMutex{},
 		config:      cfg,
 		handlers:    make(map[string]handler.Handler),
-		futureJobs:  make(map[string]time.Time),
+		futureJobs:  make(map[string]*jobs.Job),
 		cron:        cron.New(),
 		cancelFuncs: []context.CancelFunc{},
 	}
@@ -322,7 +321,7 @@ func (p *PgBackend) Enqueue(ctx context.Context, job *jobs.Job) (jobID string, e
 	// add future jobs to the future job list
 	if job.RunAfter.After(time.Now().UTC()) {
 		p.mu.Lock()
-		p.futureJobs[jobID] = job.RunAfter
+		p.futureJobs[jobID] = job
 		p.mu.Unlock()
 		p.logger.Debug("added job to future jobs list", "queue", job.Queue, "job_id", jobID, "run_after", job.RunAfter)
 	}
@@ -511,7 +510,7 @@ func (p *PgBackend) updateJob(ctx context.Context, jobErr error) (err error) {
 
 	if time.Until(runAfter) > 0 {
 		p.mu.Lock()
-		p.futureJobs[fmt.Sprint(job.ID)] = runAfter
+		p.futureJobs[fmt.Sprint(job.ID)] = job
 		p.mu.Unlock()
 	}
 
@@ -579,14 +578,16 @@ func (p *PgBackend) initFutureJobs(ctx context.Context, queue string) (err error
 		return
 	}
 
-	var id string
-	var runAfter time.Time
-	_, err = pgx.ForEachRow(rows, []any{&id, &runAfter}, func() error {
+	futureJobs, err := pgx.CollectRows(rows, pgx.RowToAddrOfStructByName[jobs.Job])
+	if err != nil {
+		return
+	}
+
+	for _, job := range futureJobs {
 		p.mu.Lock()
-		p.futureJobs[id] = runAfter
+		p.futureJobs[fmt.Sprintf("%d", job.ID)] = job
 		p.mu.Unlock()
-		return nil
-	})
+	}
 
 	return
 }
@@ -604,15 +605,15 @@ func (p *PgBackend) scheduleFutureJobs(ctx context.Context, queue string) {
 	for {
 		// loop over list of future jobs, scheduling goroutines to wait for jobs that are due within the next 30 seconds
 		p.mu.Lock()
-		for jobID, runAfter := range p.futureJobs {
-			timeUntillRunAfter := time.Until(runAfter)
+		for jobID, job := range p.futureJobs {
+			timeUntillRunAfter := time.Until(job.RunAfter)
 			if timeUntillRunAfter <= p.config.FutureJobWindow {
 				delete(p.futureJobs, jobID)
-				go func(jid string) {
+				go func(jid string, j *jobs.Job) {
 					scheduleCh := time.After(timeUntillRunAfter)
 					<-scheduleCh
-					p.announceJob(ctx, queue, jid)
-				}(jobID)
+					p.announceJob(ctx, j.Queue, jid)
+				}(jobID, job)
 			}
 		}
 		p.mu.Unlock()
@@ -708,7 +709,7 @@ func (p *PgBackend) handleJob(ctx context.Context, jobID string, h handler.Handl
 	}
 	defer func(ctx context.Context) { _ = tx.Rollback(ctx) }(ctx) // rollback has no effect if the transaction has been committed
 
-	job, err = p.getPendingJob(ctx, tx, jobID)
+	job, err = p.getJob(ctx, tx, jobID)
 	if err != nil {
 		return
 	}
@@ -815,8 +816,8 @@ func (p *PgBackend) release(ctx context.Context, conn *pgxpool.Conn, queue strin
 	conn.Release()
 }
 
-func (p *PgBackend) getPendingJob(ctx context.Context, tx pgx.Tx, jobID string) (job *jobs.Job, err error) {
-	row, err := tx.Query(ctx, PendingJobQuery, jobID)
+func (p *PgBackend) getJob(ctx context.Context, tx pgx.Tx, jobID string) (job *jobs.Job, err error) {
+	row, err := tx.Query(ctx, JobQuery, jobID)
 	if err != nil {
 		return
 	}

--- a/backends/postgres/postgres_backend_test.go
+++ b/backends/postgres/postgres_backend_test.go
@@ -192,8 +192,8 @@ func TestMultipleProcessors(t *testing.T) {
 
 	// From one of the neoq clients, enqueue several jobs. At least one per processor registered above.
 	nq := neos[0]
+	wg.Add(ConcurrentWorkers)
 	for i := 0; i < ConcurrentWorkers; i++ {
-		wg.Add(1)
 		ctx := context.Background()
 		deadline := time.Now().UTC().Add(10 * time.Second)
 		jid, e := nq.Enqueue(ctx, &jobs.Job{
@@ -498,6 +498,7 @@ results_loop:
 }
 
 // Test_MoveJobsToDeadQueue tests that when a job's MaxRetries is reached, that the job is moved ot the dead queue successfully
+// https://github.com/acaloiaro/neoq/issues/98
 func Test_MoveJobsToDeadQueue(t *testing.T) {
 	connString, conn := prepareAndCleanupDB(t)
 	const queue = "testing"
@@ -637,5 +638,103 @@ func TestJobEnqueuedSeparately(t *testing.T) {
 	}
 	if err != nil {
 		t.Error(err)
+	}
+}
+
+// TestBasicJobMultipleQueueWithError tests that the postgres backend is able to process jobs on multiple queues
+// and retries occur
+// https://github.com/acaloiaro/neoq/issues/98
+// nolint: gocyclo
+func TestBasicJobMultipleQueueWithError(t *testing.T) {
+	connString, conn := prepareAndCleanupDB(t)
+	const queue = "testing"
+	const queue2 = "testing2"
+
+	ctx := context.TODO()
+	nq, err := neoq.New(ctx,
+		neoq.WithBackend(postgres.Backend),
+		neoq.WithLogLevel(logging.LogLevelDebug),
+		postgres.WithConnectionString(connString))
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer nq.Shutdown(ctx)
+
+	h := handler.New(queue, func(_ context.Context) (err error) {
+		return
+	})
+
+	h2 := handler.New(queue2, func(_ context.Context) (err error) {
+		panic("no good")
+	})
+
+	err = nq.Start(ctx, h)
+	if err != nil {
+		t.Error(err)
+	}
+
+	err = nq.Start(ctx, h2)
+	if err != nil {
+		t.Error(err)
+	}
+
+	job2Chan := make(chan string, 100)
+	go func() {
+		jid, err := nq.Enqueue(ctx, &jobs.Job{
+			Queue: queue,
+			Payload: map[string]interface{}{
+				"message": fmt.Sprintf("should not fail: %d", internal.RandInt(10000000000)),
+			},
+		})
+		if err != nil || jid == jobs.DuplicateJobID {
+			t.Error(err)
+		}
+
+		maxRetries := 1
+		jid2, err := nq.Enqueue(ctx, &jobs.Job{
+			Queue: queue2,
+			Payload: map[string]interface{}{
+				"message": fmt.Sprintf("should fail: %d", internal.RandInt(10000000000)),
+			},
+			MaxRetries: &maxRetries,
+		})
+		if err != nil || jid2 == jobs.DuplicateJobID {
+			t.Error(err)
+		}
+
+		job2Chan <- jid2
+	}()
+
+	// wait for the job to process before waiting for updates
+	jid2 := <-job2Chan
+
+	// ensure job has fields set correctly
+	maxWait := time.Now().Add(30 * time.Second)
+	var status string
+	for {
+		if time.Now().After(maxWait) {
+			break
+		}
+
+		err = conn.
+			QueryRow(context.Background(), "SELECT status FROM neoq_dead_jobs WHERE id = $1", jid2).
+			Scan(&status)
+
+		if err == nil {
+			break
+		}
+
+		// jid2 is empty until the job gets queued
+		if jid2 == "" || err != nil && errors.Is(err, pgx.ErrNoRows) {
+			time.Sleep(100 * time.Millisecond)
+			continue
+		} else if err != nil {
+			t.Error(err)
+			return
+		}
+	}
+
+	if status != internal.JobStatusFailed {
+		t.Error("should be dead")
 	}
 }

